### ISSUE TITLE
Race condition when pulling+running named containers

### ISSUE THREAD

**Issue by jpetazzo** [docker/swarm#427]
**What I do:** 

On a clean cluster (no image present):

`docker run --name lemon -d busybox sleep 42 & docker  run --name lemon -d busybox sleep 42`

**What I expect to see:**

The ID of one container, and an error message about the name `lemon` being already in use.

**What I see instead:**

Two containers are started, confirmed by:

```
$ docker ps
CONTAINER ID        IMAGE                       COMMAND             CREATED             STATUS                  PORTS               NAMES
3f3f9c68f531        busybox:buildroot-2014.02   "sleep 42"          6 seconds ago       Up Less than a second                       52.11.76.162/lemon   
ca85e952ee6a        busybox:buildroot-2014.02   "sleep 42"          15 seconds ago      Up Less than a second                       52.11.85.52/lemon    
```

I believe Swarm checks the uniqueness of the name at the wrong time (i.e. before carrying on with the pull).

Ideas:
- having a global "name lock manager" in Swarm, so that before starting the pull, Swarm locks the name (maybe for a given time, so that if the machine doing the pull burns in flames, the name doesn't stay locked forever...)
- when doing the run, and when receiving events, use the lock manager, to report things like "WARNING: a container with an existing duplicate name was started!"

```
$ docker version
Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3.3
Git commit (client): 5bc2ff8
OS/Arch (client): linux/amd64
Server version: swarm/0.1.0
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 2acbea1
```

**Comment by aluzzardi** [comment]
This is way worse than that - the entire scheduler has this race condition I believe.

If you `docker run -m 10g`, swarm blocks pulling, and right after you `docker run -m 10g`, you might end up scheduling 2 `10g` containers on the same machine (that happens only to have `10g` of ram).

**Comment by abronan** [comment]
@aluzzardi Agreed, this one is trickier than it seems. Just throwing an idea but we could probably put each request and their attached configuration into a queue at the Manager level. We could pick a node and pull images in parallel for each entry in the queue but only `run` containers once they dequeue. This gives some room to verify resource constraints and naming collisions as well. If at the `run` phase we see that we violate a resource constraint, we can move the request at the end of the queue and restart the process of picking a node. We lose in parallelism but improve in correctness.

For the **naming collision** part I feel like using a `name lock manager` should be the answer too but as of now I would put it into the Key/Value store which is still under heavy thinking and can be subject to a lot of changes. We could still put a local cache on the Manager from the `name` entries on each `RequestedState` in the KV but this would be temporary as we need this `name lock manager` to be replicated on other Masters as well..

Still thinking about this though.

**Comment by aluzzardi** [comment]
Indeed, this one is tricky. We've been thinking for a while on how to fix it and we basically reached the same conclusions you just did.

This seems to be dependent on the key/value store.

I'm not sure if we should postpone this until we actually have a K/V around? What do you think?

**Comment by vieux** [comment]
@aluzzardi @abronan what about checking if the node can still carry the container after the pull, and if not retry the whole process ?

Surely it doesn't fix the issue completely, but it helps https://github.com/docker/swarm/pull/481

**Comment by abronan** [comment]
@vieux Yes, sounds like a good idea to circumvent the issue until we find a better way. We could check for the name as well after the pulling process and throw an error if a container already exists with that name (even though we still need to embed this information in the K/V).

**Comment by vieux** [comment]
@abronan we could add a default always on "unique name filter"
